main/lsp/requests/completion.cc

```diff
@@ -92,36 +94,94 @@ string methodSnippet(const core::GlobalState &gs, core::SymbolRef method) {
-        int prevNewlineLoc = preDefinition.rfind('\n', lastNewlineLoc - 1);
-        // if there is no '\n', we're at the top of the file, so just return documentation.
-        if (prevNewlineLoc == preDefinition.npos) {
+    std::vector<string_view> documentation_lines;
+
+    // Iterate from the last line, to the first line
+    for (auto it = all_lines.rbegin(); it != all_lines.rend(); it++) {
+        string_view line = absl::StripAsciiWhitespace(*it);
+
+        // Short circuit when line is empty
+        if (line.empty()) {
+            break;
+        }
+
+        // Handle single-line sig block
+        else if (absl::StartsWith(line, "sig")) {
+            // Do nothing for a one-line sig block
+        }
+
+        // Handle multi-line sig block
+        else if (absl::StartsWith(line, "end")) {
+            // ASSUMPTION: We either hit the start of file, a `sig do` or an `end`
+            do
+                it++;
+            while (it != all_lines.rend()
+                   // Start of sig block
+                   && !absl::StartsWith(line, "sig do")
+                   // Invalid end keyword
+                   && !absl::StartsWith(line, "end"));
+
+            // We have either
+            // 1) Reached the start of the file
+            // 2) Found a `sig do`
+            // 3) Found an invalid end keyword
+            if (it == all_lines.rend() || absl::StartsWith(line, "end")) {
+                break;
+            } else {
+                // Reached a sig block.
+                ENFORCE(absl::StartsWith(line, "sig do"));
+
+                // Stop looking if this is a single-line block e.g `sig do; <block>; end`
+                string_view after_sig_do = line.substr(strlen("sig do"));
+                after_sig_do = absl::StripAsciiWhitespace(after_sig_do);
+                if (absl::StartsWith(after_sig_do, ";") && absl::EndsWith(after_sig_do, "end")) {
+                    break;
+                }
+                // Else, this is a valid sig block. Move on to any possible documentation.
+            }
+        }
+
+        // Handle a comment line. Do not count typing declarations.
+        else if (absl::StartsWith(line, "#") && !absl::StartsWith(line, "# typed:")) {
+            // Account for whitespace before comment e.g
+            // # abc -> "abc"
+            // #abc -> "abc"
+            int skip_after_hash = absl::StartsWith(line, "# ") ? 2 : 1;
+
+            documentation_lines.push_back(line.substr(line.find('#') + skip_after_hash, line.size()));
+        }
+
+        // No other cases applied to this line, so stop looking.
+        else {
             break;
         }
-        prevLine = preDefinition.substr(prevNewlineLoc, lastNewlineLoc - prevNewlineLoc);
-        lastNewlineLoc = prevNewlineLoc;
     }
-    if (documentation.empty()) {
+
+    string documentation = absl::StrJoin(documentation_lines.rbegin(), documentation_lines.rend(), "\n");
```

Comment:
I think it's safe to trim whitespace from the end of the string; WDYT? (absl has a `StripTrailingAsciiWhitespace` method.)

Reply:
Probably won't have to because every line is stripped, but we could throw it in for good measure 👍

Reply:
On the topic of adding a newline for non-empty documentation, would it be advisable to use `StripTrailingAsciiWhitespace` to strip newlines? The name and [documentation](https://github.com/abseil/abseil-cpp/blob/master/absl/strings/ascii.h#L223) don't mention it, but the abseil [tests use it](https://github.com/abseil/abseil-cpp/blob/master/absl/strings/strip_test.cc#L178).

Reply:
If you `StrJoin` a set of empty lines, you get a string with multiple newlines -- so it can happen even though you strip whitespace from individual lines already.

Reply:
> would it be advisable to use StripTrailingAsciiWhitespace to strip newlines?

Yes, I expect the function to remove all whitespace. Newlines are whitespace too. :)

Reply:
:O cool, this will save me quite a few lines. f4db4d9ac